### PR TITLE
Wire O_DIRECT also to Uncached I/O

### LINKED ISSUE #17027: systemd-sysext fails to install extensions with Direct I/O
### System information
Type | Version/Name
 --- | ---
Distribution Name	|Fedora Linux
Distribution Version	|41 (Server Edition)
Kernel Version	|6.11.4-301.fc41.x86_64
Architecture	|x86_64
OpenZFS Version	|zfs-2.3.99-175_g390f6c119, zfs-kmod-2.3.99-175_g390f6c119

### Describe the problem you're observing
Squashfs uses a loop device to read data via its AIO handler, which sets the `O_DIRECT` flag on the request. In `zfs_setup_direct()`, the conditions [here](https://github.com/openzfs/zfs/blob/master/module/zfs/zfs_vnops.c#L244) do not account for `direct=standard` and `ioflag & O_DIRECT`. This causes most of the requests to fail due to [unalignment with the page boundary](https://github.com/openzfs/zfs/blob/master/module/zfs/zfs_vnops.c#L272).

While the [Linux block layer typically rejects requests](https://github.com/truenas/linux/blob/truenas/linux-6.12/block/fops.c#L374) that are not aligned to the block device's logical block size, [the loop device bypasses the block layer](https://github.com/truenas/linux/blob/truenas/linux-6.12/drivers/block/loop.c#L470) by directly invoking the file system's read/write handler without enforcing alignment checks. As a result, `systemd-sysext` fails reliably when attempting to install extensions on a ZFS file system.

### Describe how to reproduce the problem
The following script reliably reproduces the issue by attempting to install a simple **Hello World** extension on a ZFS file system:
``` sh
truncate -s 64M /tmp/f1
zpool create tank /tmp/f1 -O mountpoint=/root/custom_extensions
mkdir -p /root/custom_extensions/myext/usr/local/bin
echo -e '#!/bin/bash\nexec /usr/bin/echo "Hello World"' > /root/custom_extensions/myext/usr/local/bin/hello
chmod 755 /root/custom_extensions/myext/usr/local/bin/hello
mkdir -p /root/custom_extensions/myext/usr/lib/extension-release.d
echo -e "ID=_any\n" > /root/custom_extensions/myext/usr/lib/extension-release.d/extension-release.hello
mksquashfs /root/custom_extensions/myext /root/custom_extensions/hello.raw
mkdir -p /run/extensions
ln -s /root/custom_extensions/hello.raw /run/extensions/
systemd-sysext refresh
```
**Expected Result**:
The extension should install successfully, and the `systemd-sysext refresh` command should complete without errors.

**Actual Result**:
`systemd-sysext refresh` fails with error: `Failed to read metadata for image hello: Invalid argument`.

### Include any warning/errors/backtraces from the system logs
```
[ 1073.145411] loop0: detected capacity change from 0 to 8
[ 1073.159480] invalid error, dev loop0, sector 0 op 0x0:(READ) flags 0x800 phys_seg 1 prio class 0
[ 1073.171444] SQUASHFS error: Failed to read block 0x0: -22
[ 1073.174262] unable to read squashfs_super_block
```

### PR DESCRIPTION
**Pull request by amotin**
Before Direct I/O was implemented, I've implemented lighter version I called Uncached I/O.  It uses normal DMU/ARC data path with some optimizations, but evicts data from caches as soon as possible and reasonable.  Originally I wired it only to a primarycache property, but now completing the integration all the way up to the VFS.

While Direct I/O has the lowest possible memory bandwidth usage, it also has a significant number of limitations.  It require I/Os to be page aligned, does not allow speculative prefetch, etc.  The Uncached I/O does not have those limitations, but instead require additional memory copy, though still one less than regular cached I/O.  As such it should fill the gap in between.  Considering this I've disabled annoying EINVAL errors on misaligned requests, adding a tunable for those who wants to test their applications.

To pass the information between the layers I had to change a number of APIs.  But as side effect upper layers can now control not only the caching, but also speculative prefetch.  I haven't wired it to VFS yet, since it require looking on some OS specifics.

Fixes #17027

### How Has This Been Tested?
Basic read/write tests with and without O_DIRECT, observing proper cache behavior.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by amotin** [comment]
The change is quite invasive on DMU/DBUF APIs.  In some places it was not obvious what is better or where to stop, so I am open to comments on whether we'd like to change more, while already there, or less, to keep some parts of compatibility if that is even possible at all with so big change.

**Comment by tonyhutter** [comment]
Would it makes sense to activate this when the user does a `fnctl(fd, F_SET_RW_HINT, RWH_WRITE_LIFE_SHORT)`?  I believe we could key off the the file's `inode->i_write_hint` value on the kernel side.

`RWF_UNCACHED` also sounds similar to this concept, but I don't think it's merged into the kernel yet.

**Comment by amotin** [comment]
@tonyhutter I am not familiar with that fcntl, but sounds possible.  I was thinking about `posix_fadvise()` or something like it.

**Comment by amotin** [comment]
`F_SET_RW_HINT` and `inode->i_write_hint` seem about on-disk life time, not saying anything about caching.  May be they could be used for some allocation policies to reduce pool fragmentation, but not for this PR.

File: module/zfs/zfs_vnops.c [hunk 1]
```diff
@@ -243,46 +249,54 @@ zfs_setup_direct(struct znode *zp, zfs_uio_t *uio, zfs_uio_rw_t rw,
...
-	if (!zfs_dio_enabled || os->os_direct == ZFS_DIRECT_DISABLED ||
-	    zn_has_cached_data(zp, zfs_uio_offset(uio),
-	    zfs_uio_offset(uio) + zfs_uio_resid(uio) - 1)) {
-		/*
-		 * Direct I/O is disabled or the region is mmap'ed. In either
-		 * case the I/O request will just directed through the ARC.
-		 */
-		ioflag &= ~O_DIRECT;
+	if (os->os_direct == ZFS_DIRECT_ALWAYS) {
+		/* Force either direct of uncached I/O. */
```
Comment (tonyhutter) [inline module/zfs/zfs_vnops.c:253]: "direct or uncached"

**Comment by bwatkinson** [comment]
So, I am not particularly in favor of a strict being disabled by default. I think we already murky the waters enough when a write I/O is `PAGE_SIZE` aligned, but not `recordsize` aligned and we just issue it through the ARC with `direct=standard`. The strict alignment should be observed in general. Users asked for direct I/O with `O_DIRECT` and they should get an `EINVAL` returned living by the only requirement we request, which is `PAGE_SIZE` alignment.

My concern is people are expecting one thing and get another transparently. This leads to confusion and makes ZFS seem like it is not doing what they strictly asked it to do (do not make any copies of my data).

I also don't like the idea that prefetching can happen if a user explicitly requested `O_DIRECT` for data blocks. When a user informs us of their intent with `O_DIRECT` they are seriously saying (or should be), they have no desire to have FS caching take place for data. If we disregard this, then prefetching becomes a nightmare for reads.

I think if we decided to add another dataset property setting for `direct` such as maybe `relaxed`, that might make sense with this? However, the default should still be `standard` in my mind. Truth be told, we only added `direct=always` for Lustre.

I think we don't need to complicate things more in my opinion. This should be an opt in feature if anything with maybe a new dataset value for `direct`.

**Comment by amotin** [comment]
> I think we already murky the waters enough when a write I/O is `PAGE_SIZE` aligned, but not `recordsize` aligned and we just issue it through the ARC with `direct=standard`. 

My patch actually improves this case, evicting the data from DBUF and ARC caches as soon as possible, reducing cache effects as user asked.

> The strict alignment should be observed in general. Users asked for direct I/O with `O_DIRECT` and they should get an `EINVAL` returned living by the only requirement we request, which is `PAGE_SIZE` alignment.

I don't think it has to be strict other than for software testing. The man page on Linux says: "The handling of misaligned O_DIRECT I/Os also varies; they can either fail with EINVAL or fall back to buffered I/O.", so relaxed behavior is not a violation.

> My concern is people are expecting one thing and get another transparently. This leads to confusion and makes ZFS seem like it is not doing what they strictly asked it to do (do not make any copies of my data).

We've already found several examples of software having no idea about alignment, but using O_DIRECT, including so general as systemd extensions.  Considering they are "broken", I bet more typical Linux file systems to do not enforce alignment.  And for us being strict just increase suffering.  I am not interested to fix all software in the world, but happy to provide them a testing tool in a shape of a tunable.

> I also don't like the idea that prefetching can happen if a user explicitly requested `O_DIRECT` for data blocks. When a user informs us of their intent with `O_DIRECT` they are seriously saying (or should be), they have no desire to have FS caching take place for data. If we disregard this, then prefetching becomes a nightmare for reads.

The data prefetch will only activate on misaligned I/Os.  So obviously user already does not know what he is doing.  It is trivial to actually disable it now, if you insist.  But I considered that somebody might use O_DIRECT with Direct I/O disabled via module parameter, just to reduce cache trashing on some parts of workload, and prefetch really give performance improvement in many cases even with NVMe pools if application queue depth is insufficient.

> I think if we decided to add another dataset property setting for `direct` such as maybe `relaxed`, that might make sense with this? However, the default should still be `standard` in my mind. Truth be told, we only added `direct=always` for Lustre.

I was actually thinking about additional `relaxed` value for `direct` property, but present code asserts on any unknown property value, so addition of any new one would be a pain with a new pool features added, etc.   I personally could live with `zfs_dio_strict=1` by default, overriding it locally, but I can't agree that it is productive for the mass market.

**Comment by adamdmoss** [comment]
Quick question:

>  Originally I wired it only to a primarycache property

Do you intend to keep this wired to `primarycache!=all` in addition to to directio-is-misaligned paths?

**Comment by amotin** [comment]
> > Originally I wired it only to a primarycache property
> 
> Do you intend to keep this wired to `primarycache != all` in addition to the directIO-is-misaligned paths?

Yes.  Exactly the same effect is expected from either of `primarycache=metadata`, `O_DIRECT` + `zfs_dio_enabled=0`, `direct=always` + `zfs_dio_enabled=0`, `O_DIRECT` + misaligned buffers, or `direct=always` + misaligned buffers.

**Comment by amotin** [comment]
Just fixed the commend typo and rebased.

**Comment by robn** [comment]
I was thinking about this today and then realised I’d misunderstood exactly what `direct=standard` did (and so is why #17027 surprised me). I thought `direct=standard` was "try direct, fall back otherwise", not caring about alignment, but I think this it what we have (and tell me if I’m wrong):

|            | aligned | unaligned | `O_DIRECT`+aligned | `O_DIRECT`+unaligned |
| ---------- | ------- | --------- | ------------------ | -------------------- |
| `disabled` | ARC     | ARC       | ARC                | ARC                  |
| `standard` | ARC     | ARC       | direct             | `EINVAL`             |
| `always`   | direct  | ARC       | direct             | ARC                  |

So it seems there’s two variables here:

- what to do if a **non-direct** request with **correct alignment** for direct arrives (column 1)
- what to do if a **direct** request with **incorrect alignment** for direct arrives (column 4)

If they were both options, `direct=` might have the following options:

|                  | aligned | unaligned | `O_DIRECT`+aligned | `O_DIRECT`+unaligned |
| ---------------- | ------- | --------- | ------------------ | -------------------- |
| `disabled`       | ARC     | ARC       | ARC                | ARC                  |
| `relaxed`        | ARC     | ARC       | direct             | ARC                  |
| `strict`         | ARC     | ARC       | direct             | `EINVAL`             |
| `always+relaxed` | direct  | ARC       | direct             | ARC                  |
| `always+strict`  | direct  | ARC       | direct             | `EINVAL`             |

So if I’m understanding this PR, the new `zfs_dio_strict` tunable effectively toggles the meaning of `standard` between `relaxed` and `strict`, right?

---

With my sysadmin hat on,  I think I prefer `relaxed` by default, as a kind of principle of least surprise.

> Users asked for direct I/O with O_DIRECT and they should get an EINVAL returned living by the only requirement we request, which is PAGE_SIZE alignment.

It feels like there's two things going on here.

One is that the _user_ didn't ask for `O_DIRECT`, but the application did. If it fails, is it going to be clear to the user what to do next? Can they find out that the problem is that the call is misaligned? Do they even have an option, like, what if the software isn't configurable?

The other is about "the only requirement we request". Until we added support for `STATX_DIOALIGN`, there wasn't even a way for an application to discover this, and even then, it's still a relatively recent facility. Without that, we have no way of communicating alignment requirements to applications.

It seems unsurprising that there are `O_DIRECT`-using software doing it "wrong". By forcing strict, all we're doing is punishing the users of those applications, possibly with no recourse. For the ones doing it "right", none of this affects them.

So it seems like "work, but slower" is a kinder thing to do. Not to mention that that's historically what OpenZFS has always done with `O_DIRECT`.

---

I don’t know if we actually want all the property values as described above (if we even could change the property values we currently have without causing problems, though I have an idea for that). But at least, I think switching standard over to “relaxed” is probably the right thing to do.

If we do want more visibility, we could have a kstat counter noting the number of times a direct IO request redirected to the ARC for service due to misalignment. It’s not super obvious of course, but combined with a system call trace on the process it at least gives the operator the ability to see what’s going on.

**Comment by tonyhutter** [comment]
>> I don't think it has to be strict other than for software testing. The man page on Linux says: "The handling of misaligned O_DIRECT I/Os also varies; they can either fail with EINVAL or fall back to buffered I/O.", so relaxed behavior is not a violation.

> So it seems like "work, but slower" is a kinder thing to do. Not to mention that that's historically what OpenZFS has always done with O_DIRECT.

I agree with @amotin and @robn that relaxed should be the default.  It works the best "out of the box" with the real world.  Swaying my opinion is fact that users can now use `statx(STATX_DIOALIGN)` to get the alignment, so the `EINVAL` feedback on non-page alignment is less needed.

**Comment by amotin** [comment]
Enabled Direct and Uncached I/O in edge cases of receive.  Since normal receive bypasses ARC, its edge cases, such as unexpected large blocks and spill blocks, should not be cached either.  Prefetch we should not need there too.

File: include/sys/dmu.h [hunk 2]
```diff
@@ -532,6 +532,26 @@ void dmu_redact(objset_t *os, uint64_t object, uint64_t offset, uint64_t size,
...
+	DMU_READ_PREFETCH	= 0,	  /* Try speculative prefetch. */
+	DMU_READ_NO_PREFETCH	= 1 << 3, /* Don't prefetch speculatively. */
+	DB_RF_NOPREFETCH	= DMU_READ_NO_PREFETCH,
+	DMU_READ_NO_DECRYPT	= 1 << 4, /* Don't decrypt. */
+	DB_RF_NO_DECRYPT	= DMU_READ_NO_DECRYPT,
+	DMU_DIRECTIO		= 1 << 5, /* Bypass ARC. */
+	DMU_UNCACHEDIO		= 1 << 6, /* Reduce caching. */
+	DMU_PARTIAL_FIRST	= 1 << 7, /* First partial access. */
+	DMU_PARTIAL_MORE	= 1 << 8, /* Following partial access. */
+	DMU_KEEP_CACHING	= 1 << 9, /* Don't affect caching. */
```
Comment (robn) [inline include/sys/dmu.h:552]: It took me a while to understand the meaning of the cache flags, and (if I'm right), I think we could name them a bit more clearly. What do you think of:

```
DMU_DIRECTIO     -> DMU_BYPASS_CACHE
DMU_UNCACHEDIO   -> DMU_DONT_CACHE
DMU_KEEP_CACHING -> DMU_RETAIN_CACHE (or `DMU_KEEP_CACHE`)
```

("keep caching" kept making me think of "keep on truckin'", like we're encouraging the cache in these difficult times: "hang in there, cache, you got this". Which says more about my mental state than anything else, maybe).

File: include/sys/dmu.h [hunk 3]
```diff
@@ -532,6 +532,26 @@ void dmu_redact(objset_t *os, uint64_t object, uint64_t offset, uint64_t size,
 void dmu_write_policy(objset_t *os, dnode_t *dn, int level, int wp,
     struct zio_prop *zp);
 
+/*
+ * DB_RF_* are to be used for dbuf_read() or in limited other cases.
```
Comment (robn) [inline include/sys/dmu.h:536]: To make it easier to ready, would it be better to either have a separate set of `DBUF_READ_` flags with just the ones that can work, or, have `dbuf_read()` (etc) take `DMU_READ_` flags but then mask off the valid ones?

(again, nbd, it just took me a while to see it. Not least, I didn't even see that the `DB_` and `DMU_` prefixes were different until I start writing this comment :grimacing:)

File: module/zfs/dmu.c [hunk 4]
```diff
@@ -1275,10 +1253,17 @@ dmu_write_impl(dmu_buf_t **dbp, int numbufs, uint64_t offset, uint64_t size,
...
-		else
-			dmu_buf_will_dirty(db, tx);
+		if (tocpy == db->db_size) {
+			dmu_buf_will_fill_flags(db, tx, B_FALSE, flags);
+		} else {
+			if (i == numbufs - 1 && bufoff + tocpy < db->db_size) {
+				if (bufoff == 0)
+					flags |= DMU_PARTIAL_FIRST;
+				else
+					flags |= DMU_PARTIAL_MORE;
```
Comment (robn) [inline module/zfs/dmu.c:1263]: Does `flags` need to be reset when we loop?

**Comment by robn** [comment]
Initial read done, a few comments added. Nothing structurally weird jumps out of me. I don't have the brain this evening to think very hard about the details, but it looks like most of this is just threading flags through and setting some of the new dbuf fields when necessary, so at worst, it's going to be one missed or wrong (which might be bad, but y'know).

It would have been a lot easier if it were separate commits, first updating the API calls and threading them all through, then moving where some things are set around, and then the semi-unrelated stuff like the direct IO changes. But at this point I'm past the tricky bit, so nbd now :)

File: include/sys/dmu.h (hunk 3 above)
Comment (amotin) [reply include/sys/dmu.h:536]: Originally those two sets of flags were independent, declared in different files. But there were several cases where dbuf flags were passed to dmu functions.  Allowing more dmu functions to receive dmu flags meant that I either have to pass two sets of flags, or somehow unify them.  I decided the second, which means that to avoid conflicts they should better be defined in one place.  But same time I am not sure I like unifying the naming, since not all dmu flags are applicable to dbufs, and not all dbuf flags are applicable to dmu.

File: module/zfs/dmu.c (hunk 4 above)
Comment (amotin) [reply module/zfs/dmu.c:1263]: This and similar parts are executed only on the last loop iteration, so nothing cares about the flags after.  The only exception is `dmu_write_uio_dnode()`, but there I do save original flags and reset if needed.

File: include/sys/dmu.h (hunk 3 above)
Comment (robn) [reply include/sys/dmu.h:536]: Yeah, I get it - it's not really an easy thing to solve in C. In another language, I'd have two separate enum-style types, and some sort of implicit conversion. It can be done with macros, but because it can't really be enforced it's always going to be a little shaky. I'm cool with leaving it there.

File: include/sys/dmu.h (hunk 2 above)
Comment (robn) [reply include/sys/dmu.h:552]: As a naive reader, I might read "direct IO" or "uncached IO" and I wonder if those flags are only for use with those particular IO types or have some special meaning in those cases. On the other hand, more generic terms may mean that you can know less about the overall system to understand them, but might make me harder to understand a broader feature because now you have to know how these pieces fit together.

And yes, I misunderstood the meaning of the last one.

I guess all up, I don't really feel like I love these names, but I don't have any better suggestions, and they probably don't really make it more difficult to understand what's happening. I have asked, and that's the important thing :)

File: module/zfs/dmu.c (hunk 4 above)
Comment (robn) [reply module/zfs/dmu.c:1263]: Ahh so you do, nice.

**Review by robn — APPROVED** [review]
Alright, this works for me. I think the feature is worthwhile, and the changes seem to make sense.

I'm not sure I have the confidence to say yes, it's definitely right and safe, but getting it onto the master branch now is worth doing. Once it lands, I'll pull it onto my daily drivers at least; not that I do a ton of `O_DIRECT`, but at least it's getting a workout.

With the `dio_strict` change (addition), I do think this would be better done with a property, but as we've discussed, we don't quite have that as an option right now. I will say that I'm quietly working on a "versioned properties" feature, that would make it possible to safely "upgrade" a property to a new version with different values. I'm not suggesting this wait for that, but it will definitely be a good test case and there's time to get that in before 2.4 if we want it.

**Comment by amotin** [comment]
Thinking about @bwatkinson request of separate accounting for Uncached I/O, I've decided to pass `DMU_UNCACHEDIO` into `zfs_racct_read()` when called from `arc_read()`.  In case of `primarycache=metadata` it might not match what we pass to `zfs_racct_write()` from DMU, but it is the best we have here.  Though I am not sure I want to expose it as a stats, creating possible confusions.

**Review by behlendorf — APPROVED** [review]
This lighter weight version of O_DIRECT is a nice way to cover that middle ground between the full Direct IO path and the buffered path.  As for changing the default behavior to be "relaxed", I agree, it's the practical thing to do given the amount of software we've seen have problems.

> I was actually thinking about additional relaxed value for direct property, but present code asserts on any unknown property value

I think adding the additional values `relaxed` and `always_relaxed` for this property would be best.  We don't need to include that in this PR, but we can start laying the ground work by including a patch in the 2.3.3 release to tolerate the new unknown values.  We should be able to leverage the 'iuv' semantics ("ignore unknown value") added in 5405be036516280c00d481bc44394d8ef7ab15c6 to make this a bit easier.

A fairly carefully walked through all the mechanical changes to the APIs while reviewing this. I didn't spot anything wrong, but it'd be easy to overlook something.  We'll want to keep an eye out for any unintentional changes in behavior when this merges.

The changes the `dmu_*` interfaces and going to badly break the Lustre builds against ZFS.  I briefly looked at providing wrapper functions, but in this case I'm not sure it's worth it.  We can detect the API changes on the Lustre side easily enough and make the needed changes.  It would actually be a good time to more carefully re-review Lustres usage of the DMU API since it's evolved a fair bit since the Lustre code was originally integrated with it.

Lastly I went ahead and resubmitted the failing tests again.  Since this change is so large I'd like to make sure we get a clean run before merging it.  Particularly with ztest since it's always good at catching corner cases missed by the test suite.

File: man/man4/zfs.4 [hunk 5]
```diff
@@ -304,14 +304,19 @@ Default dnode block size as a power of 2.
 .It Sy zfs_default_ibs Ns = Ns Sy 17 Po 128 KiB Pc Pq int
 Default dnode indirect block size as a power of 2.
 .
-.It Sy zfs_dio_enabled Ns = Ns Sy 0 Ns | Ns 1 Pq int
+.It Sy zfs_dio_enabled Ns = Ns Sy 1 Ns | Ns 0 Pq int
```
[the remark this answers is not in the document]
Comment (amotin) [reply man/man4/zfs.4:307]: I hope we'll fix it at some point.  Though this change reduces the difference.
